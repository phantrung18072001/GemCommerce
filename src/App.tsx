import { useState } from 'react';
import { InputNumber, RadioGroup } from './components';

const UNIT = {
  PERCENTAGE: 'PERCENTAGE',
  PIXEL: 'PIXEL',
};

const App = () => {
  const [unit, setUnit] = useState(UNIT.PERCENTAGE);
  const [value, setValue] = useState(0);

  const options = [
    { label: '%', value: UNIT.PERCENTAGE },
    { label: 'px', value: UNIT.PIXEL },
  ];

  return (
    <div className="w-screen h-screen bg-neutral-950 flex items-center justify-center text-neutral-100">
      <div className="w-70 bg-[#151515] p-4 rounded-lg">
        <div className="space-y-4">
          <div className="flex gap-2 items-center ">
            <div className="text-sm w-25 text-[#AAAAAA]">Unit</div>
            <RadioGroup
              name="unit"
              options={options}
              defaultValue={unit}
              className="h-9 w-35"
              onChange={(value) => {
                setUnit(value);
              }}
            />
          </div>
          <div className="flex gap-2 items-center">
            <div className="text-sm w-25 text-[#AAAAAA]">Value</div>
            <InputNumber
              className="h-9 w-35"
              min={0}
              value={unit === UNIT.PERCENTAGE && value > 100 ? 100 : value}
              max={unit === UNIT.PERCENTAGE ? 100 : undefined}
              onChange={(value) => {
                setValue(value);
              }}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default App;
