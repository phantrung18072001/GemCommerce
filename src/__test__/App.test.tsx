import { describe, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import App from '../App';
import userEvent from '@testing-library/user-event';

let container: HTMLElement;

describe('App Component', () => {
  beforeEach(() => {
    ({ container } = render(<App />));
  });

  it('renders correctly', async () => {
    await waitFor(() => {
      expect(container).toMatchSnapshot();
      expect(screen.getByText('Unit')).toBeInTheDocument();
      expect(screen.getByText('Value')).toBeInTheDocument();
    });
  });

  describe('Test Unit Field', () => {
    it('should display both % and px unit options', () => {
      const percentageRadio = screen.getByDisplayValue('PERCENTAGE');
      const pixelRadio = screen.getByDisplayValue('PIXEL');

      expect(percentageRadio).toBeInTheDocument();
      expect(pixelRadio).toBeInTheDocument();

      // Check labels
      expect(screen.getByText('%')).toBeInTheDocument();
      expect(screen.getByText('px')).toBeInTheDocument();
    });

    it('should allow switching between % and px units', async () => {
      const percentageRadio = screen.getByDisplayValue('PERCENTAGE');
      const pixelRadio = screen.getByDisplayValue('PIXEL');

      // % by default
      expect(percentageRadio).toBeChecked();
      expect(pixelRadio).not.toBeChecked();

      // Switch to px
      await userEvent.click(pixelRadio);
      expect(pixelRadio).toBeChecked();
      expect(percentageRadio).not.toBeChecked();

      // Switch back to %
      await userEvent.click(percentageRadio);
      expect(percentageRadio).toBeChecked();
      expect(pixelRadio).not.toBeChecked();
    });
  });

  describe('Test Value Field', () => {
    it('should allow integer and float values', async () => {
      const inputElement = screen.getByTestId('number-input');
      await userEvent.type(inputElement, '12');
      fireEvent.blur(inputElement);
      expect(inputElement).toHaveValue('12');

      await userEvent.clear(inputElement);
      await userEvent.type(inputElement, '12.5');
      fireEvent.blur(inputElement);
      expect(inputElement).toHaveValue('12.5');
    });

    it('should replace comma with dot', async () => {
      const inputElement = screen.getByTestId('number-input');
      await userEvent.type(inputElement, '12,3');
      fireEvent.blur(inputElement);
      expect(inputElement).toHaveValue('12.3');
    });

    it('should remove non-numeric characters on blur', async () => {
      const pixelRadio = screen.getByDisplayValue('PIXEL');
      await userEvent.click(pixelRadio);
      const testCases = [
        { testInput: '123a', expected: '123' },
        { testInput: '12a3', expected: '12' },
        { testInput: 'a123', expected: '123' },
        { testInput: '12.4.5', expected: '12.4' },
      ];
      const inputElement = screen.getByTestId('number-input');

      for (const { testInput, expected } of testCases) {
        await userEvent.clear(inputElement);
        await userEvent.type(inputElement, testInput);
        fireEvent.blur(inputElement);
        expect(inputElement).toHaveValue(expected);
      }
    });

    it('should set value to 0 when input is less than 0', async () => {
      const input = screen.getByTestId('number-input');
      await userEvent.type(input, '-5');
      fireEvent.blur(input);
      expect(input).toHaveValue('0');
    });
  });

  describe('Percentage Unit Test', () => {
    it('should disable decrease button when value is 0', async () => {
      const input = screen.getByTestId('number-input');
      const decreaseBtn = screen.getByTestId('decrease-btn');

      await userEvent.clear(input);
      await userEvent.type(input, '0');
      fireEvent.blur(input);

      expect(decreaseBtn).toBeDisabled();
    });

    it('should update value to 100 when switching from px to % and current value > 100', async () => {
      const pixelRadio = screen.getByDisplayValue('PIXEL');
      await userEvent.click(pixelRadio);

      const input = screen.getByTestId('number-input');
      await userEvent.clear(input);
      await userEvent.type(input, '200');
      fireEvent.blur(input);
      expect(input).toHaveValue('200');

      const percentageRadio = screen.getByDisplayValue('PERCENTAGE');
      await userEvent.click(percentageRadio);

      expect(input).toHaveValue('100');
    });
  });
});
