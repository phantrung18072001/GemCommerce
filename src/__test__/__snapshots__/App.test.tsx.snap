// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > renders correctly 1`] = `
<div>
  <div
    class="w-screen h-screen bg-neutral-950 flex items-center justify-center text-neutral-100"
  >
    <div
      class="w-70 bg-[#151515] p-4 rounded-lg"
    >
      <div
        class="space-y-4"
      >
        <div
          class="flex gap-2 items-center "
        >
          <div
            class="text-sm w-25 text-[#AAAAAA]"
          >
            Unit
          </div>
          <div
            class="flex bg-[#212121] rounded-lg p-0.5 h-9 w-35"
          >
            <label
              class="rounded-md w-full h-full text-sm flex items-center justify-center cursor-pointer bg-[#424242] text-neutral-100"
            >
              <input
                checked=""
                class="hidden"
                name="unit"
                type="radio"
                value="PERCENTAGE"
              />
              <span>
                %
              </span>
            </label>
            <label
              class="rounded-md w-full h-full text-sm flex items-center justify-center cursor-pointer text-neutral-400 hover:text-neutral-200"
            >
              <input
                class="hidden"
                name="unit"
                type="radio"
                value="PIXEL"
              />
              <span>
                px
              </span>
            </label>
          </div>
        </div>
        <div
          class="flex gap-2 items-center"
        >
          <div
            class="text-sm w-25 text-[#AAAAAA]"
          >
            Value
          </div>
          <div
            class="flex items-center justify-center rounded-md bg-[#212121]
        has-[input:hover]:bg-[#424242] has-[input:focus]:border-[#3c67ff] has-[input:focus]:border-1 has-[input:focus]:border-solid h-9 w-35
      "
          >
            <div
              class="relative"
            >
              <button
                class="h-9 w-9 cursor-pointer rounded-tl-md rounded-bl-md hover:bg-[#424242] disabled:opacity-50 disabled:cursor-not-allowed"
                data-testid="decrease-btn"
                disabled=""
              >
                -
              </button>
            </div>
            <div
              class="flex-1 bg-transparent"
            >
              <input
                class="w-full border-none outline-none text-center text-sm"
                data-testid="number-input"
                name="number-input"
                type="text"
                value="0"
              />
            </div>
            <button
              class="h-9 w-9 cursor-pointer rounded-tr-md rounded-br-md hover:bg-[#424242] disabled:opacity-50 disabled:cursor-not-allowed"
              data-testid="increase-btn"
            >
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
